### PR TITLE
Add parameter event filter class

### PR DESCRIPTION
**Pull request by tfoote**
This came out of review of https://github.com/ros2/rclcpp/pull/371#discussion_r146718046

### CODE REVIEW

File: rclcpp/include/rclcpp/parameter_client.hpp [hunk 1]
```diff
@@ -268,6 +268,66 @@ class SyncParametersClient
 };
 
 }  // namespace parameter_client
+
+
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/parameter_client.hpp]: nitpick: minimize vertical whitespace

File: rclcpp/include/rclcpp/parameter_client.hpp [hunk 2]
```diff
@@ -268,6 +268,66 @@ class SyncParametersClient
...
+    }
+  }
+
+  /// Get the result of the filter
+  /**
+   * \return A std::vector<EventPair> of all matching parameter changes in this event.
+   */
+  RCLCPP_PUBLIC
+  inline
+  const std::vector<EventPair> & list()
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/parameter_client.hpp]: Shouldn't this be a const method too?

File: rclcpp/include/rclcpp/parameter_client.hpp [hunk 3]
```diff
@@ -268,6 +268,66 @@ class SyncParametersClient
...
+class ParameterEventsFilter
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(ParameterEventsFilter)
+  enum class EventType {NEW, DELETED, CHANGED};  ///< An enum for the type of event.
+  /// Used for the listed results
+  using EventPair = std::pair<EventType, rcl_interfaces::msg::Parameter *>;
+
+  RCLCPP_PUBLIC
+  inline ParameterEventsFilter(
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/parameter_client.hpp]: I believe the use of `inline` here (and below) is redundant:

> A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function.

-- http://en.cppreference.com/w/cpp/language/inline

File: rclcpp/include/rclcpp/parameter_client.hpp (hunk 2 above)
Comment (wjwwood) [reply rclcpp/include/rclcpp/parameter_client.hpp]: Maybe a more descriptive name? Something like `get_events` or `get_filtered_events`.

File: rclcpp/test/test_parameter_events_filter.cpp [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+    p.name = "deleted";
+    p.value.integer_value = 1;
+    full->deleted_parameters.push_back(p);
+    dp->changed_parameters.push_back(p);
+  }
+
+  // void TearDown()
+  // {
+  // }
```
Comment (wjwwood) [inline rclcpp/test/test_parameter_events_filter.cpp]: remove this?

File: rclcpp/test/test_parameter_events_filter.cpp [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  rclcpp::ParameterEventsFilter::EventType dt = rclcpp::ParameterEventsFilter::EventType::DELETED;
+};
+
+/*
+   Testing filters.
+ */
+TEST_F(TestParameterEventFilter, full_by_type) {
+  auto res = rclcpp::ParameterEventsFilter(full,
+      {"new", "changed", "deleted"},
+      {nt, ct, dt});
```
Comment (wjwwood) [inline rclcpp/test/test_parameter_events_filter.cpp]: nitpick: the indentation here looks like it is one level too many

Comment (tfoote) [reply rclcpp/test/test_parameter_events_filter.cpp]: I thought so too, but uncrustify insisted I push it in 4 not 2.

21: --- test/test_parameter_events_filter.cpp
21: +++ test/test_parameter_events_filter.cpp.uncrustify
21: @@ -77,2 +77,2 @@
21: -    {"new", "changed", "deleted"},
21: -    {nt, ct, dt});
21: +      {"new", "changed", "deleted"},
21: +      {nt, ct, dt});
21:

Comment (dirk-thomas) [reply rclcpp/test/test_parameter_events_filter.cpp]: The problem is that the code doesn't wrap immediately after the open parenthesis. If you follow that the indentation should be one level only.

**Comment by tfoote** [comment]
Rebased after #371 merged and cleaned up implementation and added documenatation.

Linux: [Build Status](http://ci.ros2.org/job/ci_linux/3560/)
Linux aarch64: [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/758/)
OSX: [Build Status](http://ci.ros2.org/job/ci_osx/2900/)
Windows: [Build Status](http://ci.ros2.org/job/ci_windows/3645/)

**Comment by tfoote** [comment]
@ros2/team please review

File: rclcpp/include/rclcpp/parameter_events_filter.hpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(ParameterEventsFilter)
+  enum class EventType {NEW, DELETED, CHANGED};  ///< An enum for the type of event.
+  /// Used for the listed results
+  using EventPair = std::pair<EventType, rcl_interfaces::msg::Parameter *>;
+
+  /// Construct a filtered view of a parameter event.
+  /**
+   * \param[in] event The parameter event message to filter.
+   * \param[in] names A list of parameter names of interest.
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/parameter_events_filter.hpp]: This is only a suggestion, but having been away from this code for a while it took me a minute to understand again what the names argument and types arguments were for, so maybe a slightly more in depth explanation of their purpose, perhaps with a simple example, in the body of the doc block would be nice I think.

File: rclcpp/src/rclcpp/time_source.cpp [hunk 7]
```diff
@@ -167,39 +169,28 @@ void TimeSource::clock_cb(const builtin_interfaces::msg::Time::SharedPtr msg)
...
-        bool value = new_parameter.value.bool_value;
-        if (value) {
-          parameter_state_ = SET_TRUE;
-          enable_ros_time();
-        } else {
-          parameter_state_ = SET_FALSE;
-          disable_ros_time();
-        }
-      }
+  rclcpp::ParameterEventsFilter filter(event, {"use_sim_time"},
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/time_source.cpp]: This API is a bit unwieldy, which is not a bad thing (at least I can't think of a better way to do it), it's just not obvious at first glance, maybe a comment here to say something like "filter for new or changed events of use_sim_time" would help to make it more obvious when someone is reading this code.

File: rclcpp/src/rclcpp/time_source.cpp [hunk 8]
```diff
@@ -167,39 +169,28 @@ void TimeSource::clock_cb(const builtin_interfaces::msg::Time::SharedPtr msg)
...
+      disable_ros_time();
     }
   }
-  for (auto & deleted_parameter : event->deleted_parameters) {
-    if (deleted_parameter.name == "use_sim_time") {
-      // If the parameter is deleted mark it as unset but dont' change state.
-      parameter_state_ = UNSET;
-    }
+  rclcpp::ParameterEventsFilter deleted(event, {"use_sim_time"},
+    {rclcpp::ParameterEventsFilter::EventType::DELETED});
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/time_source.cpp]: Here it could be "filter everything but the deletion of the use_sim_time parameter".

**Pull request merged**